packages/server-wallet/src/utilities/messaging.ts

```diff
@@ -19,43 +19,34 @@ export function mergeOutgoing(outgoing: Notice[]): Notice[] {
 
   const {sender} = outgoing[0].params;
 
-  const mergedOutgoing: Record<string, Payload> = {};
-
-  for (const notice of outgoing) {
-    const {recipient, data} = notice.params as {recipient: string; data: Payload};
-    if (!mergedOutgoing[recipient]) {
-      mergedOutgoing[recipient] = {walletVersion: WALLET_VERSION};
+  const messages = outgoing.map(o => o.params as WireMessage);
+
+  return _.map(
+    _.groupBy(messages, o => o.recipient),
+    (rcptMsgs, recipient) => {
+      const states = uniqueAndSorted(rcptMsgs.flatMap(n => n.data.signedStates || []));
+      const requests = uniqueAndSorted(rcptMsgs.flatMap(n => n.data.requests || []));
+      const objectives = uniqueAndSorted(rcptMsgs.flatMap(n => n.data.objectives || []));
+
+      return {
+        method: 'MessageQueued' as const,
+        params: {
+          sender,
+          recipient,
+          data: {
+            walletVersion: WALLET_VERSION,
+            signedStates: states.length > 0 ? states : undefined,
+            requests: requests.length > 0 ? requests : undefined,
+            objectives: objectives.length > 0 ? objectives : undefined,
+          },
+        },
+      };
     }
-
-    const {signedStates, requests, objectives} = mergedOutgoing[recipient];
-
-    mergedOutgoing[recipient] = {
-      walletVersion: WALLET_VERSION,
-      signedStates: mergeProp(signedStates, data.signedStates),
-      requests: mergeProp(requests, data.requests),
-      objectives: mergeProp(objectives, data.objectives),
-    };
-  }
-  return Object.keys(mergedOutgoing).map(k => ({
-    params: {sender, recipient: k, data: mergedOutgoing[k]},
-    method: 'MessageQueued' as const,
-  }));
+  );
 }
 
-function mergeProp<T>(a: T[] | undefined, b: T[] | undefined): T[] | undefined {
-  if (a && b) {
-    return _.orderBy(
-      _.uniqWith(_.concat(a, b), _.isEqual),
-      ['channelId', 'turnNum'],
-      ['desc', 'asc']
-    );
-  } else if (a) {
-    return a;
-  } else if (b) {
-    return b;
-  } else {
-    return undefined;
-  }
+function uniqueAndSorted<T>(array: T[]): T[] {
```

Comment:
Considering that this commit is about removing `O(N^2)` behaviour, can you leave a comment about the complexity of this function?